dotnet-interactive/Lsp/LspSerializer.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Newtonsoft.Json;
+using Newtonsoft.Json.Converters;
+using Newtonsoft.Json.Linq;
+using Newtonsoft.Json.Serialization;
+
+namespace Microsoft.DotNet.Interactive.App.Lsp
+{
+    public static class LspSerializer
+    {
+        public static readonly JsonSerializerSettings JsonSerializerSettings;
+
+        public static JsonSerializer JsonSerializer { get; }
+
+        static LspSerializer()
+        {
+            JsonSerializerSettings = CreateDefaultSettings();
+            JsonSerializer = JsonSerializer.Create(JsonSerializerSettings);
+        }
+
+        private static JsonSerializerSettings CreateDefaultSettings()
+        {
+            var settings = new JsonSerializerSettings
+            {
+                ContractResolver = new CamelCasePropertyNamesContractResolver(),
+                Formatting = Newtonsoft.Json.Formatting.None,
+                MissingMemberHandling = MissingMemberHandling.Error,
+                NullValueHandling = NullValueHandling.Ignore,
+                ReferenceLoopHandling = ReferenceLoopHandling.Ignore,
+            };
+
+            settings.Converters.Add(new StringEnumConverter(new CamelCaseNamingStrategy()));
+            return settings;
+        }
+
+        public static bool TryToObject<T>(this JToken token, out T result)
+        {
+            // TODO: is there a faster way to do this?
+            var foundError = false;
+            var settings = CreateDefaultSettings();
+            settings.Error = (sender, e) =>
+            {
+                foundError = true;
+                e.ErrorContext.Handled = true;
+            };
+            var serializer = JsonSerializer.Create(settings);
```

Comment:
why building this if you have the static properties?

Reply:
I need to add the `.Error` handler, but only when this `TryToObject<T>()` method is called, so I can't modify the static value or else it'll pollute all other (de)serialization calls.  I couldn't think of another way to have local-only error swallowing without `catch (JsonSerializationException)`.

Reply:
Are you trying to implement partial success? Why are errors occurring?

---
In reply to: [409909923](https://github.com/dotnet/interactive/pull/348#discussion_r409909923) [](ancestors = 409909923)

Reply:
I'm trying to prevent leaking internal implementation details if the user submits bad or wrong JSON.  If they payload isn't correct, the deserialization will throw and that seemed sloppy, but ultimately I don't feel too strongly about it.

Reply:
I am talking about the fact that we keep rebuilding the serialization settings and the serializer using the exact same code that the type initializer is using to create the static property JsonSerializer  and the static field JsonSerializerSettings. should we reuse those objects isntead of keep on building them at each serialization?

Reply:
There are two things happening here.  The static settings (`.JsonSerializerSettings`) and static serializer (`.JsonSerializer`) are populated once in the static constructor, and they're regularly reused and never recreated.

I think the part you're concerned about is the `.TryToObject<T>()` method, and that's because we need to know, only in this function call, if there was a deserialization error we need to mark the error as handled to prevent a `JsonSerializationException` from being thrown so that we can instead return a `bool` value.  I can't set the `.Error` property on the static settings object because then any subsequent call to the regular static serializer/settings will have the `.Error` handler set, and that's not what's wanted.

The alternative would be to use the static values and instead make the function body look like this:

``` C#
public static bool TryToObject<T>(this JToken token, out T result)
{
    try
    {
        result = token.ToObject<T>(JsonSerializer);
        return true;
    }
    catch (JsonSerializationException)
    {
        result = default(T);
        return false;
    }
}
```

I opted away from this because throwing/catching exceptions is expensive and @jonsequitur was concerned about serialization/deserialization perf.

Reply:
As per Jon's other comments and what we discussed this morning, this will change to throw/404/500/etc. in case of a malformed request, because this is a situation where the person doing the wrong thing is either us or somebody calling our API and _not_ the end user.